libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py

```diff
@@ -1828,6 +1828,19 @@ def write_UART_config(self, f):
             return
         f.write('\n// UART configuration\n')
 
+        # write out which serial ports we actually have
+        idx = 0
+        nports = 0
+        serial_order = self.get_config('SERIAL_ORDER', required=False, aslist=True)
+        for serial in serial_order:
+            if serial == 'EMPTY':
+                f.write('#define HAL_HAVE_SERIAL%u 0\n' % idx)
+            else:
+                f.write('#define HAL_HAVE_SERIAL%u 1\n' % idx)
+                nports = nports + 1
+            idx += 1
+        f.write('#define HAL_NUM_SERIAL_PORTS %u\n' % nports)
```

Comment:
Does this conflict with the one further down?

https://github.com/ArduPilot/ardupilot/pull/25090/files#diff-00cdd2b90e14e9bac5a19c02a223046a45d1eb38242337320621df067739b84cL1958

Reply:
No, two different numbers - the second should really be called "highest serial port index" as it is used to size arrays etc. The one I am writing is the *actual* number of serial ports